screensaverd.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Simple screen saver daemon. Turns off the display after idle timeout.
+ * Turns the display on on any event from /dev/input/event* files.
+ * On exit turns the display on.
+ *
+ * Copyright (C) 2015 Jolla Ltd.
+ * Contact: Igor Zhbanov <igor.zhbanov@jolla.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define _BSD_SOURCE
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <dirent.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <sys/select.h>
+
+#define UNUSED			__attribute__((unused))
+
+#define DISPLAY_CONTROL		"/sys/class/graphics/fb0/blank"
+#define DEV_INPUT_DIR		"/dev/input"
+#define LOG_NAME		"screensaverd"
+#define EVENT_PREFIX		"event"
+#define MAX_DEVICES		32
+#define DISPLAY_OFF_TIME	25 /* seconds */
+#define BUF_SIZE		4096
+
+sig_atomic_t volatile running = 1;
+
+/* ------------------------------------------------------------------------ */
+
+/* perror() with printf-style format and application name prefix. */
+static void
+perrorf(const char *format, ...)
+{
+	va_list ap;
+
+	va_start(ap, format);
+	printf("[%s] ", LOG_NAME);
+	vprintf(format, ap);
+	printf(": %s\n", strerror(errno));
+	va_end(ap);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static void
+close_fds(int fds[], int num)
+{
+	int i;
+
+	for (i = 0; i < num; i++)
+		close(fds[i]);
+}
+
+/* ------------------------------------------------------------------------ */
+
+/* Open all /dev/input/event* files. */
+static int
+open_fds(int fds[], int *num, int max_num)
+{
+	DIR *dir;
+	struct dirent *d;
+
+	if (!(dir = opendir(DEV_INPUT_DIR))) {
+		perrorf("Can't open directory %s", DEV_INPUT_DIR);
+		return -1;
+	}
+
+	while ((d = readdir(dir))) {
+		char name[256];
+
+		if (strncmp(d->d_name, EVENT_PREFIX, strlen(EVENT_PREFIX)))
+			continue; /* Not /dev/input/event* file */
+
+		snprintf(name, sizeof(name), "%s/%s", DEV_INPUT_DIR,
+			 d->d_name);
+		if ((fds[*num] = open(name, O_RDONLY)) == -1) {
+			perrorf("Can't open input device %s", name);
+			close_fds(fds, *num);
+			closedir(dir);
+			return -1;
```

Comment:
It is more the design decision than technical. We are running with root credentials but for some reason cannot open the /dev/input/event-something. Why is that? Does our kernel broken?

We can continue here, but I don't know whether it is good.
Also what to count as error here? If we can open none of files? So should we return success if we opened only one device?

IMHO the problem here is that we are not checking what devices we are opening and what devices we need. We need, for sure, at least some buttons. (Ideally, at least power button.) But what device it is? To not hardcode its name, we are opening all event devices. But what if it is the button device we couldn't open? Then there could be a situation that we will turn the display off but never turn it on because we can't get needed events.